torch/utils/_pytree.py

```diff
@@ -89,6 +90,9 @@
 FromDumpableContextFn = Callable[[DumpableContext], Context]
 ToStrFunc = Callable[["TreeSpec", List[str]], str]
 MaybeFromStrFunc = Callable[[str], Optional[Tuple[Any, Context, str]]]
+KeyEntry = Any  # Should be Hashable
+KeyPath = Tuple[KeyEntry, ...]
```

Comment:
Can this be something more concrete? Since the original discussion was on sources, 2 quesitons:

1) Why not use sources directly?

2) And if rejecting sources directly, should we have a source like object in that it has both the base and the accessor together? So instead of DictKey being something like `DictKey("foo")`, it would be `DictKey(base="my_dict", key="foo")`

Reply:
> Why not use sources directly?

Mostly a layering thing; I consider pytree to be a lower level component that shouldn't necessarily "know" about higher-level concepts like shape env/dynamo/sources. Ideally these concepts map onto each other fairly closely.

> And if rejecting sources directly, should we have a source like object in that it has both the base and the accessor together? So instead of DictKey being something like DictKey("foo"), it would be DictKey(base="my_dict", key="foo")

Great question, let me think about that. I rather uncritically replicated the JAX API but maybe it can be improved on for our use case.

Reply:
Sources are a lot lower level than dynamo or shape env imo - they are a primitive that fulfills the invariant of `eval(source.name())` giving you back the user object referred to in the source. 

We should reuse them where possible, especially as we start chatting about the generic pt2 agnostic guards/specialization system (as we have).

Reply:
With Voz's original comment over chat, I thought that we were going to write a utility to track sources through pytree.tree_flatten. The disadvantage of the key path API is that we now need users to provide key path rules for their custom pytree types.

To avoid the layering issue: pytree.tree_flatten shouldn't use sources in its implementation. Would it be possible to do something like turn on Dynamo's source tracking and call pytree.tree_flatten to obtain the sources?

Reply:
That seems like an even worse layering violation haha; a system that wants only to use pytrees must now opt into using dynamo in order to use a relatively simple functionality. 

I will investigate the sources stuff to see if it can be pulled out of dynamo.

Reply:
Independently I think the key-path API is useful by itself, but it seems more robust for the AOTExport use case to not rely on users defining key path rules for custom pytree classes

Reply:
Yeah it's an extra step, but I think it's okay. Users need to specify serialization rules as well anyway. 

If we are really worried about this, is to implement the flat container fallback that JAX has. If a `flatten_with_keys` fn is not defined, then we when we try to flatten with paths, a special `KeyEntry` is returned that signals that we first need to call `tree_flatten` and then index into the flattened list to retrieve the value. This trades efficiency for guaranteed coverage.

In general, I prefer to be explicit vs. implicit with the export APIs, so I didn't implement it that way here, but I think it's a reasonable change to make since key paths are not export-specific.

As a related note, JAX has a nice class-based registration API that I like that we should consider implementing. That is a slightly more robust way to do registration, since it does not separate registration from class definition (and thus increase the likelihood that the class is changed but not the pytree registration functions).

Reply:
Discussed offline/chat. We agreed that if we do this, we want generic sources free of dynamo notions. As such, we would either move the guard_source,reconstruct stuff to a helper fn that visits the source types and dispatches behavior based on them OR as a mixin to provide the dynamo functions to implement, but allowing for other systems to not know about the dynamo side of the type.

Reply:
@voznesenskym bat signaled me, so here is my spicy take.

I think that strict / non-strict should have the same way of describing the structure on input types. However, in both cases, I think we should *not* use Dynamo sources. Dynamo sources work on arbitrary Python data structure, not only those that have pytree rules defined for them, and they are abstraction breaking (e.g., if a tensor is stored on a private field `_foo`, we will just slap together a source `x._foo` in this case.) In contrast, export input description is part of the public API, so you want it simple, and you need some abstraction mechanisms that allow custom data types to say how exactly they would like to present their internal structure (and not just blat out all their guts.)

I think it's probably OK for export to use keypaths. It's pretty simple and there's already a custom registration system you can lift off. But then strict mode should still use keypaths too!